modules/misc/xdg-user-dirs.nix

```
@@ -86,25 +86,41 @@ in {
       example = { XDG_MISC_DIR = "$HOME/Misc"; };
       description = "Other user directories.";
     };
+
+    createDirs = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        Whether to automatically create the XDG user directories.
+      '';
+    };
   };
 
-  config = mkIf cfg.enable {
+  config = let
+    directories = {
+      XDG_DESKTOP_DIR = cfg.desktop;
+      XDG_DOCUMENTS_DIR = cfg.documents;
+      XDG_DOWNLOAD_DIR = cfg.download;
+      XDG_MUSIC_DIR = cfg.music;
+      XDG_PICTURES_DIR = cfg.pictures;
+      XDG_PUBLICSHARE_DIR = cfg.publicShare;
+      XDG_TEMPLATES_DIR = cfg.templates;
+      XDG_VIDEOS_DIR = cfg.videos;
+    } // cfg.extraConfig;
+  in mkIf cfg.enable {
     xdg.configFile."user-dirs.dirs".text = let
-      options = {
-        XDG_DESKTOP_DIR = cfg.desktop;
-        XDG_DOCUMENTS_DIR = cfg.documents;
-        XDG_DOWNLOAD_DIR = cfg.download;
-        XDG_MUSIC_DIR = cfg.music;
-        XDG_PICTURES_DIR = cfg.pictures;
-        XDG_PUBLICSHARE_DIR = cfg.publicShare;
-        XDG_TEMPLATES_DIR = cfg.templates;
-        XDG_VIDEOS_DIR = cfg.videos;
-      } // cfg.extraConfig;
-
       # For some reason, these need to be wrapped with quotes to be valid.
-      wrapped = mapAttrs (_: value: ''"${value}"'') options;
+      wrapped = mapAttrs (_: value: ''"${value}"'') directories;
     in generators.toKeyValue { } wrapped;
 
     xdg.configFile."user-dirs.conf".text = "enabled=False";
+
+    home.activation = mkIf cfg.createDirs {
+      createXdgUserDirectories = let
+        directoriesList = attrsets.attrValues directories;
+        mkdir = (dir: ''mkdir -p "${dir}"'');
+      in lib.hm.dag.entryAfter [ "writeBoundary" ]
+      (strings.concatMapStringsSep "\n" mkdir directoriesList);
```

I think something like
```suggestion
      in lib.hm.dag.entryAfter [ "writeBoundary" ] ''
        $DRY_RUN_CMD mkdir -p $VERBOSE_ARG \
          ${concatStringsSep " \\  " directoriesList}
      '';
```
would be better (untested).

---

`escapeShellArgs` didn't work, since it escaped `$HOME` variable.

---

Returned to the previous way, but adding `$DRY_RUN_CMD` and `$VERBOSE_ARG`.

---

One way to make this works is to break backwards compatibility here. Like, changing `$HOME` to `home.homeDirectory`.

Actually, it is just half a backwards compatibility change, since the only actual issue would be if someone set `xdg-user-dirs.createDirs = true;`. For the people that doesn't, this would works well.

WDYT @rycee ?

---

Well, didn't work, got this:

```
The option value `home.homeDirectory' in `/nix/store/k2lv9nx64xm8rgy3639ldyf1x1mi0s2y-source/modules/home-environment.nix' is not of type `path'.
(use '--show-trace' to show detailed location information)
```

---

Strange that the file that I am getting the error is a completely another one. I even wrote a test to see if I can reproduce the error, in the test itself it works fine but when I try it in my config it does not.

Here is a pastebin with my changes: http://ix.io/2RBP

---

I think I understand the issue now, I can't define a submodule using the definition of a module since the module is probably evaluated later on.